### PR TITLE
docker image sprinkler with java, s3 get object

### PR DESCRIPTION
**Pull request by schowsf**
risk:low 

sprinkler docker image with java to enable command running java 

orchard runner to parse artifact for bucket and s3 location

s3 func to getObject

### CODE REVIEW

File: orchard/orchard_runner.go [hunk 1]
```diff
@@ -37,3 +38,17 @@ func parseCommandLine(command string) ([]string, error) {
...
+
+// parse artifact expecting a s3 URI to return the bucket and path
+func (c OrchardStdoutRunner) ParseArtifact(artifact string) (string, string, error) {
+	r, err := regexp.Compile(`^s3://([^ /]+)/([^ ]+)$`)
+	if err != nil {
+		return "", "", err
+	}
+	res := r.FindStringSubmatch(artifact)
+	if len(res) == 3 {
+		return res[1], res[2], nil
```
Comment (damianxu88) [inline orchard/orchard_runner.go:50]: @schowsf `res[0]` is the `artifact` passed in right?

File: object_store/simple_s3.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// ListBuckets lists the buckets in the current account.
+func (basics BucketBasics) ListBuckets() ([]types.Bucket, error) {
+	result, err := basics.S3Client.ListBuckets(context.TODO(), &s3.ListBucketsInput{})
+	var buckets []types.Bucket
+	if err != nil {
+		log.Printf("Couldn't list buckets for your account. Here's why: %v\n", err)
+	} else {
+		buckets = result.Buckets
+	}
+	return buckets, err
```
Comment (damianxu88) [inline object_store/simple_s3.go:43]: @schowsf poor man (aka go)'s version of `Either`? lol

**Review by damianxu88 — APPROVED** [review]
LGTM

File: object_store/simple_s3.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package object_store
```
Comment (realstraw) [inline object_store/simple_s3.go:1]: can put it in a more generic package, like "common" or something

File: object_store/simple_s3.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+// BucketBasics encapsulates the Amazon Simple Storage Service (Amazon S3) actions
+// used in the examples.
+// It contains S3Client, an Amazon S3 service client that is used to perform bucket
+// and object actions.
+type BucketBasics struct {
+	S3Client *s3.Client
+}
+
+func DefaultS3Client() *s3.Client {
```
Comment (realstraw) [inline object_store/simple_s3.go:24]: I think this should return the `err` too

File: data/exampleWorkflow.json [hunk 5]
```diff
@@ -0,0 +1 @@
...urce.EmrResource","resourceSpec":{"releaseLabel":"emr-6.7.0","applications":[],"serviceRole":"EMR_DefaultRole","resourceRole":"EMR_EC2_DefaultRole","tags":[],"bootstrapActions":null,"configurations":null,"instancesConfig":{"subnetId":"${?SPADE_AWS_SUBNET_ID}","instanceCount":2,"masterInstanceType":"m6g.xlarge","slaveInstanceType":"m6g.xlarge","ec2KeyName":"orchard","additionalMasterSecurityGroups":null,"additionalSlaveSecurityGroups":null}},"maxAttempt":3,"terminateAfter":null}],"dependencies":{"48838514-da9d-4334-b74f-78198ca4054d":["199e4c88-ff0b-45f3-b05f-df8c56c1b947"],"199e4c88-ff0b-45f3-b05f-df8c56c1b947":["1864bcf0-0d64-44e9-84a4-eb58faeb1f15"],"2c0a2603-b6db-45f9-9320-75a0c22ddd6d":["48838514-da9d-4334-b74f-78198ca4054d"],"1864bcf0-0d64-44e9-84a4-eb58faeb1f15":["1022ca6b-bb26-416b-b572-8b5a226ea056"]},"actions":[{"id":"SnsAlarm-d9f5fe7c-eec5-4431-ae0e-999eab583443","name":"name","actionType":"aws.action.SnsAlarm","actionSpec":{"topicArn":"topic","subject":"Failure Alarm","message":"failed"}},{"id":"SnsAlarm-fce035af-253e-40c1-9769-2e06fdfb5262","name":"name","actionType":"aws.action.SnsAlarm","actionSpec":{"topicArn":"topic","subject":"Success Alarm","message":"success"}}]}
```
[the remark this answers is not in the document]
Comment (schowsf) [reply data/exampleWorkflow.json]: done

File: go.mod [hunk 6]
```diff
@@ -12,6 +12,24 @@ require (
...
+	github.com/aws/aws-sdk-go-v2/feature/ec2/imds v1.13.0 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/configsources v1.1.30 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/endpoints/v2 v2.4.24 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/ini v1.3.31 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/v4a v1.0.22 // indirect
+	github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding v1.9.11 // indirect
+	github.com/aws/aws-sdk-go-v2/service/internal/checksum v1.1.25 // indirect
+	github.com/aws/aws-sdk-go-v2/service/internal/presigned-url v1.9.24 // indirect
+	github.com/aws/aws-sdk-go-v2/service/internal/s3shared v1.13.24 // indirect
+	github.com/aws/aws-sdk-go-v2/service/s3 v1.30.6 // indirect
```
[the remark this answers is not in the document]
Comment (schowsf) [reply go.mod]: i just learned about direct vs indirect here.  Thanks

File: object_store/simple_s3.go (hunk 3 above)
Comment (schowsf) [reply object_store/simple_s3.go]: done

File: object_store/simple_s3.go (hunk 4 above)
Comment (schowsf) [reply object_store/simple_s3.go]: done

File: docker-compose.yml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 services:
   ws:
-    image: golang
+    image: sprinkler
```
Comment (realstraw) [inline docker-compose.yml:3]: would `build: .` instead of `image` work?

File: go.mod [hunk 8]
```diff
@@ -3,55 +3,70 @@ module mce.salesforce.com/sprinkler
...
+	github.com/aws/aws-sdk-go-v2/config v1.18.18
+	github.com/aws/aws-sdk-go-v2/service/s3 v1.30.6
+	github.com/gin-gonic/gin v1.9.0
 	github.com/google/uuid v1.3.0
 	github.com/spf13/cobra v1.6.1
 	github.com/spf13/viper v1.15.0
-	gorm.io/driver/postgres v1.4.7
-	gorm.io/gorm v1.24.5
+	gorm.io/driver/postgres v1.5.0
+	gorm.io/gorm v1.24.7-0.20230306060331-85eaf9eeda11
```
Comment (realstraw) [inline go.mod:14]: is `-0.20230306060331-85eaf9eeda11` necessary?

File: go.mod [hunk 9]
```diff
@@ -3,55 +3,70 @@ module mce.salesforce.com/sprinkler
 go 1.19
 
 require (
-	github.com/gin-gonic/gin v1.8.2
+	github.com/aws/aws-sdk-go-v2 v1.17.6
+	github.com/aws/aws-sdk-go-v2/config v1.18.18
+	github.com/aws/aws-sdk-go-v2/service/s3 v1.30.6
```
[the remark this answers is not in the document]
Comment (schowsf) [reply go.mod:8]: I tried without the top 2, `go run` brought them back as require `direct`

File: docker-compose.yml (hunk 7 above)
Comment (schowsf) [reply docker-compose.yml:3]: that `build: . ` works.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,7 @@
+From openjdk:11-jre as with_java
```
[the remark this answers is not in the document]
Comment (schowsf) [reply Dockerfile:1]: I see openjdk more a place holder.  

I think our deployment would use 
`dva-registry.internal.salesforce.com/sfci/docker-images/sfdc_centos7_openjdk11/sfdc_centos7_openjdk11u4`

Though on github, the internal.salesforce may not be public avail.

[the remark this answers is not in the document]
Comment (damianxu88) [reply Dockerfile]: yes, deployment will use environment specific docker image, think the image we use here should be as generic as possible

File: common/simple_s3.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+// It contains S3Client, an Amazon S3 service client that is used to perform bucket
+// and object actions.
+type BucketBasics struct {
+	S3Client *s3.Client
+}
+
+func DefaultS3Client() (*s3.Client, error) {
+	sdkConfig, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		fmt.Println("Couldn't load default configuration. Have you set up your AWS account?")
```
[the remark this answers is not in the document]
Comment (schowsf) [reply common/simple_s3.go]: done

File: docker-compose.yml (hunk 7 above)
Comment (schowsf) [reply docker-compose.yml]: change committed

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,7 @@
+FROM openjdk:11-jre as with_java
+
+FROM golang:1.20
```
Comment (anniezhang77) [inline Dockerfile:3]: does it matter when this version is not consistent with 1.19 in go.mod?

Comment (schowsf) [reply Dockerfile:3]: both works on `go run` 

I noticed differences in the size of the docker image.  `cc3a4c313856` is golang 1.20.  `dcf1b1bad329` is golang 1.19

```
REPOSITORY         TAG           IMAGE ID       CREATED          SIZE
sprinkler_ws       latest        cc3a4c313856   2 days ago       915MB

REPOSITORY         TAG           IMAGE ID       CREATED          SIZE
sprinkler_ws       latest        dcf1b1bad329   16 seconds ago   1.13GB
```

Seems 1.20 is smaller, any objections to switching to 1.20 in the `go.mod`?

Comment (damianxu88) [reply Dockerfile:3]: @schowsf no objection my end

Comment (schowsf) [reply Dockerfile:3]: ^^ I mean mix of 1.19 and 1.20 between `Dockerfile` and `go.mod`, or consistently 1.19 or 1.20.  all combinations worked.

Comment (anniezhang77) [reply Dockerfile:3]: go for it

Comment (anniezhang77) [reply Dockerfile:3]: I'm for 1.20 for both.

Comment (realstraw) [reply Dockerfile]: let's make everything consistent and latest stable version

File: orchard/orchard_runner.go [hunk 13]
```diff
@@ -37,3 +38,17 @@ func parseCommandLine(command string) ([]string, error) {
 	err := json.Unmarshal([]byte(command), &output)
 	return output, err
 }
+
+// parse artifact expecting a s3 URI to return the bucket and path
+func (c OrchardStdoutRunner) ParseArtifact(artifact string) (string, string, error) {
```
Comment (realstraw) [inline orchard/orchard_runner.go:43]: is this called some where? I cannot seem to find it

Comment (schowsf) [reply orchard/orchard_runner.go:43]: not called yet.   I plan to use it in the next PR.

File: Dockerfile (hunk 12 above)
Comment (schowsf) [reply Dockerfile]: okay.  1.20 commit pushed.

File: orchard/orchard_runner.go (hunk 13 above)
Comment (schowsf) [reply orchard/orchard_runner.go]: I have taken it out.

Comment (realstraw) [reply orchard/orchard_runner.go]: or maybe add the next part to show a very simplified usage?

Comment (schowsf) [reply orchard/orchard_runner.go]: I have take out `ParseArtifact` in this PR.  Better to separate it in the next PR with the rest of the logic.  I plan to have a different struct `OrchardJavaRunner` then.

Comment (realstraw) [reply orchard/orchard_runner.go]: ok,  I'll merge this one this time. But it's better to try implement complete (but minimum viable) feature in PRs. Those features can be very lean with rough edges, but having unused code like `simple_s3.go` in a service repo is not desired state. Not much of the changes in this PR have any real usage other than providing utilities.

**Pull request merged**